File: parametricEQ.cpp
/*
  ==============================================================================

    parametricEQ.cpp
    Created: 16 Sep 2023 1:28:22pm
    Author:  Navaneeth Suresh Kumar

  ==============================================================================
*/

#include "parametricEQ.h"


parametricEQ::parametricEQ() {
    a0 = 0.f;
    a1 = 0.f;
    a2 = 0.f;
    b1 = 0.f;
    b2 = 0.f;
    d0 = 0.f;
    w0 = 1.f;

    int numInputs = 2;
    for (int i = 0; i < numInputs; i++) {
        a1_delay.add(0.f);
        a2_delay.add(0.f);
        b1_delay.add(0.f);
        b2_delay.add(0.f);
    }

}
parametricEQ::~parametricEQ() {}

void parametricEQ::updateCoefficients(int f_center, float Q, float gain, double fs) {
    float Pi = juce::MathConstants<float>::pi;
    float theta = 2.f * Pi * f_center / fs;
    float U = powf(10.f, (gain / 20.f));
    float E = 4.f / (1.f + U);
    float B = 0.5f * (1.f - (E * tan(theta / 2.f * Q))) / (1.f + (E * tan(theta / 2.f * Q)));
    float G = (0.5f + B) * cos(theta);

    a0 = 0.5f - B;
    a1 = 0.f;
    a2 = -(0.5f - B);
    b1 = -2.f * G;
    b2 = 2.f * B;
    w0 = U - 1.f;
    d0 = 1.f;


}
void parametricEQ::processAudio(juce::AudioBuffer<float>& buffer) {
    float z = 0.f;
    auto totalNumInputChannels = buffer.getNumChannels();
    for (int i = 0; i < totalNumInputChannels; i++) {
        auto* channelData = buffer.getWritePointer(i);
        for (int n = 0; n < buffer.getNumSamples(); n++) {
            z = (*channelData) * a0 + a1 * a1_delay[i] + a2 * a2_delay[i] - b1 * b1_delay[i] - b2 * b2_delay[i];
            a2_delay.set(i, a1_delay[i]);
            b2_delay.set(i, b1_delay[i]);
            a1_delay.set(i, *channelData);;
            b1_delay.set(i, z);
            *channelData = d0 * (*channelData) + w0 * z;
            channelData++;
        }
    }

}

void parametricEQ::resetDelay(int numInputs){
    
    
    for (int i = 0; i<numInputs; i++){
        a1_delay.set(i, 0.f);
        a2_delay.set(i, 0.f);
        b1_delay.set(i, 0.f);
        b2_delay.set(i, 0.f);
    }
    
}
void parametricEQ::updateDryWetRatio(float ratio){
    d0 = ratio;
    w0 = 1.f - ratio;
    
}

File: Matlab-code/parametricEQ.m
classdef parametricEQ
    %PARAMETRICEQ Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
          a0
        a1
        a2
        b1
        b2
        d0
        w0
        
        %delays
        a1_delay
        a2_delay
        b1_delay
        b2_delay
    end
    
    methods
        function obj = parametricEQ(f_center,Q,gain,fs)
            obj=resetDelays(obj);
            obj=updateDryWetRatio(obj,0.5);
            obj=updateCoefficients(obj, f_center,Q,gain,fs);
        end
  function obj = updateCoefficients(obj,f_center,Q,gain,fs)
            theta=2*pi*f_center/fs;
            U = 10^(gain/20);
            E= 4/(1+U);
            B=0.5*(1 - (E*tan(theta/2*Q)))/(1 + (E*tan(theta/2*Q)));
             G=(0.5+B)*cos(theta);

        obj.a0=0.5-B;
        obj.a1 = 0;
        obj.a2 = -(0.5-B);
        obj.b1 = -2*G;
        obj.b2 = 2*B;


            obj.w0 = U-1.0;
            obj.d0= 1;
         
        end
        function[out_buffer]=processAudio(obj,EQ_Fcenter,EQ_Q,EQ_gain,fs,input_buffer)
            obj=obj.updateCoefficients(EQ_Fcenter,EQ_Q,EQ_gain,fs);
           
            z = zeros(length(input_buffer),1);
            x=input_buffer;

            for n = 1:length(x)
                
                z(n) =  (x(n)*obj.a0 + ...
                    obj.a1*obj.a1_delay + obj.a2*obj.a2_delay - obj.b1*obj.b1_delay - obj.b2*obj.b2_delay);
                obj.a2_delay = obj.a1_delay;
                obj.b2_delay = obj.b1_delay;
                obj. a1_delay = x(n);
                obj.b1_delay = z(n);
            end

            out_buffer = obj.d0*x + obj.w0*z;
            
            
        end
        function obj=resetDelays(obj)
            obj.a1_delay=0;
            obj.a2_delay=0;
            obj.b1_delay=0;
            obj.b2_delay=0;

        end
        function obj= updateDryWetRatio(obj,ratio)
            obj.d0=ratio;
            obj.w0=1.0-ratio;

        end
    end
end

